### PR TITLE
[WIP] Adds wide/compact modes to CupertinoTabBar (fixes #38194)

### LINKED ISSUE #38194: CupertinoTabBar can't adaptive tablet layout
On iPad, the icons and titles of the tab bar items are arranged horizontally.  In Flutter apps they are drawn vertically (icon over title).

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, v1.8.5-pre.116, on Microsoft Windows [Version 10.0.18950.1000], locale zh-CN)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Visual Studio - develop for Windows (Visual Studio Community 2017 15.9.11)
[√] Android Studio (version 3.4)
[√] VS Code (version 1.37.0)
[√] Connected device (2 available)

• No issues found!
```

### PR DESCRIPTION
**Pull request by kerberjg**
## Description

This PR adds some functionality to CupertinoTabBar: it is now possible to make the tab bar appear with wide icons (iPad) as well as wide-compact (landscape iPhone) by setting the `isWide` and `isCompact` properties in its constructor.

A few other properties related to icon size have been added, as well as providing and adjusting the required font sizes.

## Related Issues

Fixes #38194

## Tests

Added tests:
 - Compact flag takes effect only when the wide flag is active as well

Updated tests:
- Adjusts height to account for bottom padding

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by kerberjg** [comment]
**Example implementation**

What this does, is checking whether the device is a tablet or a phone in landscape mode, and showing wide icons in those cases.

Also, in the latter case, the tab bar is also shown as "compact" (with height reduced to 32.0, as per Apple guidelines).

Resources: 

```dart
const bool isTablet = MediaQuery.of(context).size.shortestSide >= 600;

...

CupertinoApp(
  title: 'Cupertino demo',
  theme: CupertinoThemeData(),
  home: OrientationBuilder(
    builder: (context, orientation) => CupertinoTabScaffold(
      tabBar: CupertinoTabBar(
        backgroundColor: CupertinoColors.white,
        isWide: orientation == Orientation.landscape || isTablet ? true : false,
        isCompact: orientation == Orientation.landscape && isTablet ? true : false,
        items: [
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.time_solid),
            title: Text('Tab 1')
          ),
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.bookmark_solid),
            title: Text('Tab 2')
          ),
        ],
      ),

      ...

    ),
  ),
)
```

Sources: 
- https://developer.apple.com/documentation/uikit/uitabbar
- https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/custom-icons/
- https://developer.apple.com/documentation/uikit/uitabbar/1623468-itempositioning
- https://ivomynttinen.com/blog/ios-design-guidelines

**Comment by kerberjg** [comment]
Should be done, can I request someone to review the PR? (cc @LongCatIsLooong @xter)

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 1]
```diff
@@ -9,8 +9,9 @@ import 'package:flutter/widgets.dart';
 import 'colors.dart';
 import 'theme.dart';
 
-// Standard iOS 10 tab bar height.
+// Standard iOS 11 tab bar height.
 const double _kTabBarHeight = 50.0;
+const double _kTabBarCompactHeight = 32.0;
```
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: https://medium.com/@hacknicity/ipad-navigation-bar-and-toolbar-height-changes-in-ios-12-91c5766809f4

Also it's very unlikely that they're both 50pt, when comparing screenshots in native iOS apps like Photos, there is a difference in height between portrait (regular) and landscape (compact)

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 2]
```diff
@@ -118,13 +125,35 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
...
+  ///
+  /// The default value is false.
+  final bool isWide;
+
+  /// Controls whether the tab bar should be displayed in its compact mode (landscape mode on iPhone) or the regular one.
+  ///
+  /// Its value is ignored when `isWide` equals false.
+  ///
+  /// The default value is false.
+  final bool isCompact;
```
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: This is correct. I made the mistake of not noting that the behavior of Plus devices is different.

If you try the same with iPhone 6S, 7, 8 (non-Plus), they should appear in compact mode, at least they do as of iOS 12

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 3]
```diff
@@ -53,7 +54,11 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
     this.backgroundColor,
     this.activeColor,
     this.inactiveColor = CupertinoColors.inactiveGray,
-    this.iconSize = 30.0,
+    this.iconSize = 34.0,
```
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: I definitely agree that having useless parameters is something that should be avoided.

However adding a second constructor for this purpose would make relatively simple code way longer than it could be. For example this relatively simple example:

```dart
CupertinoApp(
  title: 'Cupertino demo',
  theme: CupertinoThemeData(),
  home: OrientationBuilder(
    builder: (context, orientation) => CupertinoTabScaffold(
      tabBar: CupertinoTabBar(
        backgroundColor: CupertinoColors.white,
        isWide: orientation == Orientation.landscape || isTablet ? true : false,
        isCompact: orientation == Orientation.landscape && !isTablet ? true : false,
        items: [
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.time_solid),
            title: Text('Tab 1')
          ),
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.bookmark_solid),
            title: Text('Tab 2')
          ),
        ],
      ),

      ...

    ),
  ),
)
```

...would instead become this:

```dart
CupertinoApp(
  title: 'Cupertino demo',
  theme: CupertinoThemeData(),
  home: OrientationBuilder(
    builder: (context, orientation) => {
      bool isWide = orientation == Orientation.landscape || isTablet ? true : false;
      bool isCompact = orientation == Orientation.landscape && !isTablet ? true : false;

      List<BottomNavigationBarItem> items = [
        BottomNavigationBarItem(
          icon: Icon(CupertinoIcons.time_solid),
          title: Text('Tab 1')
        ),
        BottomNavigationBarItem(
          icon: Icon(CupertinoIcons.bookmark_solid),
          title: Text('Tab 2')
        ),
      ];

      if(isCompact) {
        return CupertinoTabScaffold.compact(
          tabBar: CupertinoTabBar(
            backgroundColor: CupertinoColors.white,
            items: items
          )
        );
      } else if(isWide) {
        return CupertinoTabScaffold.wide(
          tabBar: CupertinoTabBar(
            backgroundColor: CupertinoColors.white,
            items: items
          )
        );
      } else {
        return CupertinoTabScaffold(
          tabBar: CupertinoTabBar(
            backgroundColor: CupertinoColors.white,
            items: items
          )
        );
      }

      ...

    }
  ),
)
```

I guess we can both agree that then the code just becomes unnecessary convoluted.

I'd very much rather replicate what Apple has done and UIKit, and have the tab bar automatically decide whether it should be wide, compact, or normal. As we have seen, by observing iOS's native apps, we can note when it is more correct for the tab bar to assume a specific form. I can provide more detailed notes on my observations in this area if necessary.

As a result, we'd still have 3 different parameters for icon sizes, yes, but it'd more justfied in this case, since the fidelity of the CupertinoTabBar would be improved due to the developers not having to guess that they should/can use a OrientationBuilder with a bunch of screen size calculations to make the tab bar look & feel like the native one does.

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: Yeah you're right, using different constructors to differentiate size classes is awkward. I thought we could do something like this:
```Widget Function(...) tabBar = isCompact ? CupertinoTabBar.compact : CupertinoTabBar.regular;```.
but it turns out constructor tear-offs are not supported yet. 

But I'm still not sure the additional size parameters are required. Wouldn't it also work if we give iconSize different default values based on `isWide` and `isCompact`? I.e.:

```dart
  const CupertinoTabBar({
    Key key,
    @required this.items,
    this.onTap,
    this.currentIndex = 0,
    this.backgroundColor,
    this.activeColor,
    this.inactiveColor = CupertinoColors.inactiveGray,
    double iconSize,
    this.border = const Border(
      top: BorderSide(
        color: _kDefaultTabBarBorderColor,
        width: 0.0, // One physical pixel.
        style: BorderStyle.solid,
      ),
    ),
  }) : assert(items != null),
       assert(
         items.length >= 2,
         "Tabs need at least 2 items to conform to Apple's HIG",
       ),
       assert(currentIndex != null),
       assert(0 <= currentIndex && currentIndex < items.length),
       assert(isCompact != null),
       assert(isWide != null),
       assert(!isCompact || isWide, 'isCompact cannot be true if isWide is not true)),
       this.iconSize = iconSize ?? isWide ? (isCompact ? 22 : 30) : 34,
       assert(inactiveColor != null),
       super(key: key);
```

**Comment by LongCatIsLooong** [comment]
For future reference I did some test on iOS 13 beta 5 simulators and compiled a chart. The metrics from iPad Air are very similar to iPad Pro so I omitted them. If the device has a non-zero bottom safe area inset in that device orientation, its tab bar height is notated in the form of `total tab bar height ( tab bar item height + diff)`.

The test icon apparently wasn't the best for the sake of measurement but the numbers are pretty consistent across devices and size classes:

|  Value (in pt) \ Device & size class |  iPhone Xr portrait <br>  w: C h: R  | iPhone Xr landscape <br> w: R h: C |  iPhone 8+ portrait <br>  w: C h: R  | iPhone 8+ landscape <br> w: R h: C |  iPad Pro 12.9" portrait <br>  w: R h: R  | iPad Pro 12.9" landscape <br> w: R h: R |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | 
| icon text stack layout |  vertical | horizontal | vertical | horizontal | horizontal | horizontal | 
| tab bar height | 83 (48 + 35) | 70 (48 + 22)| 49 | 49 | 65 (49 + 6) | 65 (49 + 6) |
| Icon size |  27.5 | 27.5 | 27.33 | 27.33 | 27.5 | 27.5 |
| font size |  10 | 13 | 10 | 13 | 13 | 13 |


|  Value (in pt) \ Device & size class | iPhone SE portrait <br> w: C h: R  | iPhone SE landscape <br> w: C h: C |  iPhone 8 portrait <br> w: C h: R  | iPhone 8 landscape <br> w: C h: C 
| :---: | :---: | :---: | :---: | :---: |
| icon text stack layout | vertical | horizontal | vertical | horizontal |
| tab bar height | 49 | 32 |  49 | 32 |
| Icon size | 27.5 | 21 | 27.5 | 21 |
| font size | 10 | 12 | 10 | 12 |

### tl;dr:
- the font size only shrinks when the size class is Compact x Regular.
- the icon size only shrinks when the size class is Compact x Compact.
- the stack layout only becomes vertical when the size class is Compact x Regular (i.e. font size shrinks when the stack layout becomes vertical).

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: The documentation here sounds to me `landscape mode on iPhone = compact`, which is not true on larger iPhones. Could you update it?

**Comment by kerberjg** [comment]
@LongCatIsLooong Thank you, I'll take a deeper look at this today and get back to you afterwards

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: Sure!

**Comment by KagurazakaHanabi** [comment]
Can I continue based on this?

**Comment by kerberjg** [comment]
I've been on a bit of a partial hiatus, but I'm back to work on this!

I'll go on implementing the automatic shape detection based on the parameters kindly provided by @LongCatIsLooong, and we'll see where we can go from there, feature-wise

**Comment by kerberjg** [comment]
@LongCatIsLooong Is there a default way of telling which size class the app is running on, or should I hack together a utility class for this?

**Comment by kerberjg** [comment]
I continued @LongCatIsLooong's investigation on size classes, the tabbar appearance and such, and I found some pretty interesting things.

https://docs.google.com/spreadsheets/d/13qx8B_S7RCkMPZNYAg-xTTxM-yJMgHUMd2bIcz70cN0/edit?usp=sharing

Initially I was inclined to think it'd be either tied to height class, logical height or that the physical limit would be 1000, but that would be incorrect for a bunch of reasons:

- All iPhones/iPods always have a compact height class while in landscape, yet _some of them display the tabbar as regular in landscape_
- While 11 Pro, X, Xs are 375 logical pixels tall, they still display the _tabbar as regular in landscape_
- 11 and Xr are above 400 logical pixels, but under 1000 physical pixels; _tabbar still regular in landscape_

This led me to believe that whether the navbar is compact or not, is explicitly based on the resolution, meaning that **the tabbar's height is compact when the physical pixel height is < 800** (with height being relative to device's orientation)

**Comment by kerberjg** [comment]
Should users be able to decide whether the tab bar and its items appear in a specific way, or do we force it to automatic? [It doesn't seem possible in UIKit](https://developer.apple.com/documentation/uikit/uitabbar), all the users can do is customize text/colors in one of the 3 modes (called respectively: stacked, inline, compactInline)

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 4]
```diff
@@ -165,24 +279,30 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
       );
 
     final Color inactive = CupertinoDynamicColor.resolve(inactiveColor, context);
+
+    // TODO: allow dynamic resizing
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-checking-in-comments-that-ask-questions

> TODOs should include the string TODO in all caps, followed by the GitHub username of the person with the best context about the problem referenced by the TODO in parenthesis. A TODO is not a commitment that the person referenced will fix the problem, it is intended to be the person with enough context to explain the problem. Thus, when you create a TODO, it is almost always your username that is given.
>
> Including an issue link in a TODO description is required.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 5]
```diff
@@ -165,24 +279,30 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
...
+    final bool isIconCompact = (CupertinoSizeClassHelper.getWidthSizeClass(context) == CupertinoSizeClass.compact) && (CupertinoSizeClassHelper.getHeightSizeClass(context) == CupertinoSizeClass.compact);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or--or--with-enum-values

> Avoid using if chains, ? …​ : …​, or, in general, any expressions involving enums.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 6]
```diff
@@ -233,16 +350,50 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
...
+          padding: const EdgeInsets.only(left: 4.0),
+          child: item.title,
+        )
+      );
+    }
+
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.center,
+      crossAxisAlignment: CrossAxisAlignment.center,
+      children: components
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]:
```suggestion
      children: components,
```

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 7]
```diff
@@ -272,6 +423,10 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
     Color activeColor,
     Color inactiveColor,
     Size iconSize,
+    Size wideIconSize,
+    Size compactIconSize,
+    bool isWide,
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: Should these be deleted? They seem to be unused.

File: packages/flutter/lib/src/cupertino/text_theme.dart [hunk 8]
```diff
@@ -176,6 +187,10 @@ class CupertinoTextThemeData extends Diagnosticable {
   /// The [TextStyle] of unselected tabs.
   TextStyle get tabLabelTextStyle => _tabLabelTextStyle ?? _defaults.tabLabelTextStyle;
 
+  final TextStyle _tabWideLabelTextStyle;
+  /// Typography of unselected tabs.
+  TextStyle get tabWideLabelTextStyle => _tabWideLabelTextStyle ?? _kDefaultTabWideLabelTextStyle;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/text_theme.dart]: The implementation of `_defaults` should probably be changed accordingly since a new textstyle is added.

**Review by LongCatIsLooong — COMMENTED** [review]
:+1: Thank you for putting so much effort into this! 

To your customizability question, I spoke with @HansMuller and we think it's probably better to leave them publicly configurable in the constructor. The same app can be run on android devices or even desktops so it's going to be tricky to find an automatic configuration that makes everyone happy. For the same reason we probably should reconsider introducing `CupertinoSizeClass` because `SizeClass` is only defined for iOS devices. 

Also if we introduce an automatic configuration, it's very likely a breaking change per [the new breaking change policy](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes).

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 9]
```diff
@@ -81,6 +178,18 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
   /// Must not be null.
   final List<BottomNavigationBarItem> items;
 
+  /// Controls whether the tab bar should be displayed in its compact mode (landscape mode on iPhone) or the regular one.
+  ///
+  /// The default value is [CupertinoTabBarMode.auto].
```
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: The enums have been renamed and “auto” values have been replaced with nulls. Will update the comments!

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 10]
```diff
@@ -123,13 +232,18 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
...
   /// The border of the [CupertinoTabBar].
   ///
   /// The default value is a one physical pixel top border with grey color.
   final Border border;
 
   @override
-  Size get preferredSize => const Size.fromHeight(_kTabBarHeight);
+  Size get preferredSize => _isBarCompact ? const Size.fromHeight(_kTabBarCompactHeight) : const Size.fromHeight(_kTabBarHeight);
+
+  bool get _isBarCompact => window.physicalSize.height < 800;
```
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: Can you please expand upon this?

**Comment by kerberjg** [comment]
@LongCatIsLooong What do you propose to do about `CupertinoSizeClass`? 

Also, I ran into a major roadblock in testing: as it is now, I am unable to mock the "automatic detection, landscape iPhone 8" scenario since the bar's height in build is based on `preferredSize`... which is determined by **window**'s physical height, not the one given by the MediaQuery (which can be easily mocked).
Is it possible to mock the window size somehow? I will keep looking into it, but if you have any suggestions regarding this, please do let me know

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: BTW, the previous implementation didn't specify the width either, but I can change this if you have suggestions

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (kerberjg) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: By doing so we're taking away the opportunity to change the icon sizes consistently in case someone wants to retain the automatic layout detection detection

**Comment by kerberjg** [comment]
The regression test "*Adding new tabs does not crash the app*" for #33455 fails, because it does not account for the possibility of horizontally laid out items, causing the test to fail.

During the test the CupertinoTabBar is picking its layout automatically, with a physical screen size of 0x0 pixels, leading to a decision of CxH (respectively bar and item layouts). 

```
...
════════════════════════════════════════════════════════════════════════════════════════════════════
══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
The following assertion was thrown during layout:
A RenderFlex overflowed by 57 pixels on the right.

The overflowing RenderFlex has an orientation of Axis.horizontal.
The edge of the RenderFlex that is overflowing has been marked in the rendering with a yellow and
black striped pattern. This is usually caused by the contents being too big for the RenderFlex.
Consider applying a flex factor (e.g. using an Expanded widget) to force the children of the
RenderFlex to fit within the available space instead of being sized to their natural size.
This is considered an error condition because it indicates that there is content that cannot be
seen. If the content is legitimately bigger than the available space, consider clipping it with a
ClipRect widget before putting it in the flex, or using a scrollable container rather than a Flex,
like a ListView.
The specific RenderFlex in question is: RenderFlex#6dc4b relayoutBoundary=up12 OVERFLOWING:
  creator: Row ← Padding ← _PointerListener ← Listener ← _GestureSemantics ← RawGestureDetector ←
    GestureDetector ← Semantics ← Expanded ← DefaultTextStyle ← Builder ← IconTheme ← ⋯
  parentData: offset=Offset(0.0, 0.0) (can use size)
  constraints: BoxConstraints(w=40.0, 0.0<=h<=45.0)
  size: Size(40.0, 21.0)
  direction: horizontal
  mainAxisAlignment: center
  mainAxisSize: max
  crossAxisAlignment: center
  textDirection: ltr
  verticalDirection: down
◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: [0, 0, 18]
  Actual: [0, 0, 19]
   Which: was <19> instead of <18> at location [2]

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///Users/mrgame64/Library/Developer/Flutter/packages/flutter/test/cupertino/tab_scaffold_test.dart:634:5)
<asynchronous suspension>
#5      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:124:25)
#6      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:699:19)
<asynchronous suspension>
#9      TestWidgetsFlutterBinding._runTest (package:flutter_test/src/binding.dart:679:14)
#10     AutomatedTestWidgetsFlutterBinding.runTest.<anonymous closure> (package:flutter_test/src/binding.dart:1079:24)
#16     AutomatedTestWidgetsFlutterBinding.runTest (package:flutter_test/src/binding.dart:1076:15)
#17     testWidgets.<anonymous closure> (package:flutter_test/src/widget_tester.dart:121:22)
#18     Declarer.test.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:171:27)
<asynchronous suspension>
#19     Invoker.waitForOutstandingCallbacks.<anonymous closure> (package:test_api/src/backend/invoker.dart:242:15)
#24     Invoker.waitForOutstandingCallbacks (package:test_api/src/backend/invoker.dart:239:5)
#25     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:169:33)
#30     Declarer.test.<anonymous closure> (package:test_api/src/backend/declarer.dart:168:13)
#31     Invoker._onRun.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/invoker.dart:392:25)
#45     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:384:19)
#46     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:418:5)
#47     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
(elided 28 frames from class _FakeAsync, package dart:async, package dart:async-patch, and package stack_trace)

This was caught by the test expectation on the following line:
  file:///Users/mrgame64/Library/Developer/Flutter/packages/flutter/test/cupertino/tab_scaffold_test.dart line 634
The test description was:
  Adding new tabs does not crash the app
════════════════════════════════════════════════════════════════════════════════════════════════════
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following message was thrown:
Multiple exceptions (41) were detected during the running of the current test, and at least one was
unexpected.

...
```

@LongCatIsLooong Why is the number of painted tabs specifically 18?

Also, is it okay to force the layout to represent its "classic" layout (RxV) for the purpose of this test? Or should the CupertinoTabBar's automatic layout detection code altered to default to RxV in mock scenarios?

**Comment by HansMuller** [comment]
I appreciate all of the work that you and @LongCatIsLooong have put into this PR over more than 5 months now! To get the PR landed, we're going to need to reduce it's scope per https://github.com/flutter/flutter/pull/39258#pullrequestreview-331601029.

Automatically applying tab and tab bar layout changes is a breaking change and working such a change through the system - assuming it was accepted - could take months more. Please try and find the minimal set of additional CupertinoTabBar parameters that would enable apps to configure the tab bar according to the [spreadsheet you created](https://docs.google.com/spreadsheets/d/13qx8B_S7RCkMPZNYAg-xTTxM-yJMgHUMd2bIcz70cN0/edit#gid=0).  Something like (just) an additional `compact` parameter, and a parameter to request vertical layout for the tabs. Support for applying the configuration automatically would have to be provided by a separate package for now.

**Comment by LongCatIsLooong** [comment]
If I remember it correctly all the numbers in the test were pretty much randomly chosen. But I'm not sure how did the physical screen size become 0x0, usually in a widget test its 800x600.

**Comment by LongCatIsLooong** [comment]
Also as @HansMuller pointed out, automatically determining the layout is going to be a breaking change (see [this section of the wiki](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes)). To avoid breaking existing code bases I think a good strategy would be separating the automatic layout deciding logic from this pull request, and expose the necessary parameters to allow developers to control the layout.

**Comment by LongCatIsLooong** [comment]
Hi @kerberjg, in our effort to reduce open pull requests, I'm going to close this pull request. Feel free to reopen it when you would like to start working on this again!
